azure-devops/provision-agent.ps1

```diff
@@ -0,0 +1,204 @@
+$CMakeURL = 'https://github.com/Kitware/CMake/releases/download/v3.16.4/cmake-3.16.4-win64-x64.msi'
+$LlvmURL = 'https://releases.llvm.org/9.0.0/LLVM-9.0.0-win64.exe'
+$NinjaURL = 'https://github.com/ninja-build/ninja/releases/download/v1.9.0/ninja-win.zip'
+$PythonURL = 'https://www.python.org/ftp/python/3.8.1/python-3.8.1-amd64.exe'
+$VstsAgentURL = 'https://vstsagentpackage.azureedge.net/agent/2.165.0/vsts-agent-win-x64-2.165.0.zip'
+
+$ErrorActionPreference = 'Stop'
+$ProgressPreference = 'SilentlyContinue'
+
+Function InstallVS
+{
+  Param(
+    [String]$WorkLoads,
+    [String]$Sku,
+    [String]$VSBootstrapperURL)
+  try
+  {
+    Write-Output 'Downloading VS bootstrapper ...'
+    [string]$bootstrapperExe = Join-Path ([System.IO.Path]::GetTempPath()) ([System.IO.Path]::GetRandomFileName() + '.exe')
+    curl.exe -L -o $bootstrapperExe $VSBootstrapperURL
+
+    $Arguments = ('/c', $bootstrapperExe, $WorkLoads, '--quiet', '--norestart', '--wait', '--nocache')
+
+    Write-Output "Starting Install: $Arguments"
+    $proc = Start-Process -FilePath cmd.exe -ArgumentList $Arguments -Wait -PassThru
+    $exitCode = $proc.ExitCode
+
+    if ($exitCode -eq 0 -or $exitCode -eq 3010)
+    {
+        Write-Output "Installation successful! $exitCode"
+    }
+    else
+    {
+        Write-Output "Nonzero exit code returned by the installation process : $exitCode."
+        exit $exitCode
+    }
+  }
+  catch
+  {
+    Write-Output 'Failed to install Visual Studio!'
+    Write-Output $_.Exception.Message
+    exit 1
+  }
+}
+
+Function InstallMSI
+{
+    Param(
+        [String]$Name,
+        [String]$Uri
+    )
+
+    try
+    {
+        Write-Output "Downloading $Name..."
+        [string]$tempRoot = [System.IO.Path]::GetTempPath();
+        [string]$randomRoot = Join-Path $tempRoot ([System.IO.Path]::GetRandomFileName())
+        [string]$msiPath = $randomRoot + '.msi'
+        [string]$msiExecPath = 'msiexec.exe'
+        $args = @('/i', $msiPath, '/norestart', '/quiet', '/qn')
+
+        curl.exe -L -o $msiPath $Uri
+        Write-Output "Installing $Name..."
+        $proc = Start-Process -FilePath $msiExecPath -ArgumentList $args -Wait -PassThru
+        $exitCode = $proc.ExitCode
+        if ($exitCode -eq 0 -or $exitCode -eq 3010)
+        {
+            Write-Output 'Installation successful!'
+        }
+        else
+        {
+            Write-Output "Nonzero exit code returned by the installation process : $exitCode."
+            exit $exitCode
+        }
+    }
+    catch
+    {
+        Write-Output "Failed to install $Name!"
+        Write-Output $_.Exception.Message
+        exit -1
+    }
+}
+
+Function InstallZip
+{
+    Param(
+        [String]$Name,
+        [String]$Uri,
+        [String]$Dir
+    )
+
+    try
+    {
+        Write-Output "Downloading $Name..."
+        [string]$randomRoot = Join-Path ([System.IO.Path]::GetTempPath()) ([System.IO.Path]::GetRandomFileName())

```

Comment:
This is inconsistent with the previous function that separately stored GetTempPath before constructing the randomRoot, ditto below.